java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthViewVector.java

```diff
@@ -0,0 +1,1434 @@
+   * @param start start position of data in buffer
+   * @param length length of data in buffer
+   * @param buffer data buffer containing the variable width element to be stored
+   *               in the vector
+   */
+  public void set(int index, int start, int length, ArrowBuf buffer) {
+    assert index >= 0;
+    fillHoles(index);
+    BitVectorHelper.setBit(validityBuffer, index);
+    byte[] data = new byte[length];
+    buffer.getBytes(start, data, 0, length);
+    setBytes(index, data, start, length);
+    lastSet = index;
+  }
+
+  /**
+   * Same as {@link #set(int, int, int, int, ArrowBuf)} except that it handles the case
+   * when index is greater than or equal to current value capacity of the
+   * vector.
+   * @param index position of the new value
+   * @param start start position of data in buffer
+   * @param length length of data in buffer
+   * @param buffer data buffer containing the variable width element to be stored
+   *               in the vector
+   */
+  public void setSafe(int index, int start, int length, ArrowBuf buffer) {
+    assert index >= 0;
+    handleSafe(index, length);
+    fillHoles(index);
+    BitVectorHelper.setBit(validityBuffer, index);
+    byte[] data = new byte[length];
+    buffer.getBytes(start, data, 0, length);
+    setBytes(index, data, start, length);
+    lastSet = index;
+  }
+
+
+  /*----------------------------------------------------------------*
+   |                                                                |
+   |                helper methods for setters                      |
+   |                                                                |
+   *----------------------------------------------------------------*/
+
+
+  protected final void fillHoles(int index) {
+    for (int i = lastSet + 1; i < index; i++) {
+      setBytes(i, EMPTY_BYTE_ARRAY, 0, EMPTY_BYTE_ARRAY.length);
+    }
+    lastSet = index - 1;
+  }
+
+  /**
+   * Get the length of the view.
+   * @param index The index of the element in the vector.
+   * @return The length of the element at the given index.
+   */
+  protected final int getLength(int index) {
+    if (index < 0 || index >= viewBuffer.capacity() / ELEMENT_SIZE) {
+      throw new IndexOutOfBoundsException("Index out of bounds: " + index);
+    }
+    if (isSet(index) == 0) {
+      return 0;
+    }
+    return viewBuffer.getInt(((long) index * ELEMENT_SIZE));
+  }
+
+  protected ArrowBuf allocateOrGetLastBuffer(BufferAllocator allocator, int length, List<ArrowBuf> dataBuffers) {
+    long dataBufferSize;
+    if (initialDataBufferSize > 0) {
+      dataBufferSize = initialDataBufferSize;
+    } else {
+      dataBufferSize = lastValueAllocationSizeInBytes;
+    }
+
+    if (dataBuffers.isEmpty() || dataBuffers.get(dataBuffers.size() - 1).capacity() -
+            dataBuffers.get(dataBuffers.size() - 1).writerIndex() < length) {
+      ArrowBuf newBuf = allocator.buffer(dataBufferSize);
+      dataBuffers.add(newBuf);
+    }
```

Comment:
Not so sure about this logic. Don't you want to reallocate existing buffers where possible? And, what if length is greater than dataBufferSize?

Reply:
> And, what if length is greater than dataBufferSize?

Correct we need to check the possible size with the length and allocate the adequate. But since, the dataBuffer can get more allocations, it would be better to allocate whatever the highest (`length` or `initialDataBufferSize` or `lastValueAllocationSizeInBytes`)

Reply:
>  Don't you want to reallocate existing buffers where possible? 

About position, I am merely checking how much space has been consumed in the last appeneded databuffer in the dataBuffers. Could you please explain a bit if I have understood this wrong?